File: e_sausage_maximization.py
import sys

# sys.setrecursionlimit(10**6)
from collections import Counter, defaultdict, deque
from bisect import bisect_left, bisect_right
from copy import deepcopy
import math

# input = input
input = sys.stdin.readline


class TrieNode:
    def __init__(self):
        self.child = [None, None]


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, num):
        curr = self.root
        for bit in range(46, -1, -1):
            b = (num >> bit) & 1
            if not curr.child[b]:
                curr.child[b] = TrieNode()
            curr = curr.child[b]

    def query(self, num):
        node = self.root
        if not node:
            return 0
        ans = 0
        for bit in range(46, -1, -1):
            b = (num >> bit) & 1
            tog = 1 ^ b
            if node.child[tog]:
                ans |= 1 << bit
                node = node.child[tog]
            else:
                node = node.child[b]
        return ans


def solve():
    n = int(input().strip())
    a = list(map(int, input().split()))

    pre = [0] * (n + 1)
    suf = [0] * (n + 2)

    for i in range(1, n + 1):
        pre[i] = pre[i - 1] ^ a[i - 1]

    for i in range(n, 0, -1):
        suf[i] = suf[i + 1] ^ a[i - 1]

    trie = Trie()
    plea = 0
    trie.insert(pre[0])

    for j in range(1, n + 2):
        p = trie.query(suf[j])
        plea = max(plea, p)
        if j <= n:
            trie.insert(pre[j])

    print(plea)


def main():
    t = 1
    for _ in range(t):
        solve()


if __name__ == "__main__":
    main()

File: e_sausage_maximization.cpp
#include <bits/stdc++.h>

#include <algorithm>
using namespace std;

#define loop_from(i, s, n) for (int i = s; i < n; i++)
#define loop(i, n) for (int i = 0; i < n; i++)
#define MOD 1'000'000'007
#define pb push_back()
#define pob pop_back()
#define int long long
#define nl '\n'
#define print(x) cout << (x) << endl
#define show(v)                         \
    for (auto& x : v) cout << x << " "; \
    cout << endl
#define debug(i) cout << #i << ": " << i << endl

#define infi numeric_limits<ll>::max()
#define ninfi numeric_limits<ll>::min()
#define infd numeric_limits<ld>::infinity()
#define ninfd -numeric_limits<ld>::infinity()
#define all(x) (x).begin(), (x).end()

typedef long long ll;
typedef long double ld;
typedef vector<ll> vll;

template <typename K, typename V>
using umap = unordered_map<K, V>;

static mt19937 rng(chrono::steady_clock::now().time_since_epoch().count());
#define RAND(a, b) uniform_int_distribution<long long>(a, b)(rng)
ll big = RAND(1, 1e9);
long long modPow(long long base, long long exp, long long mod) {
    long long res = 1;
    base %= mod;

    while (exp > 0) {
        if (exp % 2 == 1) {
            res = (res * base) % mod;
        }

        base = (base * base) % mod;
        exp /= 2;
    }

    return res;
}

class TrieNode {
    public:
        TrieNode* child[2];

        TrieNode() {
            child[0] = nullptr;
            child[1] = nullptr;
        }
};

class Trie {
        TrieNode root;

    public:
        Trie() { root = TrieNode(); }

        void insert(ll num) {
            TrieNode* curr = &root;
            for (int i = 46; i >= 0; i--) {
                ll b = (num >> i) & 1;
                if (!curr->child[b]) {
                    curr->child[b] = new TrieNode();
                }
                curr = curr->child[b];
            }
        }

        ll query(ll num) {
            ll ans = 0;
            TrieNode* curr = &root;

            for (int i = 46; i >= 0; i--) {
                ll b = (num >> i) & 1;
                ll tog = 1 ^ b;
                if (curr->child[tog]) {
                    ans |= ((1LL) << i);
                    curr = curr->child[tog];
                } else {
                    curr = curr->child[b];
                }
            }

            return ans;
        }
};

void solve() {
    ll n;
    cin >> n;
    vector<ll> nums(n);
    vll pre;
    ll p = 0;
    for (ll& x : nums) cin >> x, p ^= x, pre.push_back(p);
    ll s = 0;
    vll suf;
    for (int i = n - 1; i >= 0; i--) s ^= nums[i], suf.push_back(s);
    reverse(all(suf));

    Trie trie;
    ll ans = pre[n - 1];
    trie.insert(0);


    for (int i = 1; i < n; i++) {
		ans = max(ans, trie.query(suf[i]));
		ans = max(ans, pre[i]);
		trie.insert(pre[i]);
    }
    print(ans);
}

int32_t main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int t = 1;

    loop_from(_, 0, t) { solve(); }
}
